src/object/cli_obj.c

```diff
@@ -4874,54 +4630,264 @@ obj_csum_fetch(const struct dc_object *obj, daos_obj_fetch_t *args,
  */
 static int
 obj_retry_next_shard(struct dc_object *obj, struct obj_auxi_args *obj_auxi,
-		     uint64_t dkey_hash, unsigned int map_ver)
+		     unsigned int map_ver, uint32_t *shard)
 {
-	unsigned int		 next_shard, retry_size, shard_cnt, start_shard;
-	int			 rc = 0;
+	unsigned int	grp_size;
+	unsigned int	start_shard;
+	int		rc = 0;
 
 	D_WARN("Retrying replica because of %s error.\n",
 	       obj_auxi->csum_retry ? "csum" : "tx_uncertain");
 
-	rc = obj_dkey2grpmemb(obj, dkey_hash, map_ver,
-			      &start_shard, &shard_cnt);
+	/* EC retry is done by degraded fetch */
+	D_ASSERT(!obj_is_ec(obj));
+	rc = obj_dkey2grpmemb(obj, obj_auxi->dkey_hash, map_ver,
+			      &start_shard, &grp_size);
 	if (rc != 0)
 		return rc;
 
-	/* bitmap in obj_auxi_args has only 4 bytes, then let's only retry
-	 * the first 32 replicas, that should be enough for most of cases.
-	 */
-
-	retry_size = shard_cnt <= OBJ_INLINE_BTIMAP * 8 ?
-		     shard_cnt : OBJ_INLINE_BTIMAP * 8;
-	next_shard = (obj_auxi->req_tgts.ort_shard_tgts[0].st_shard + 1) %
-		     retry_size + start_shard;
-
-	if (next_shard == obj_auxi->initial_shard) {
+	*shard = (obj_auxi->req_tgts.ort_shard_tgts[0].st_shard + 1) % grp_size + start_shard;
+	if (*shard == obj_auxi->initial_shard) {
 		obj_auxi->no_retry = 1;
 		if (obj_auxi->csum_retry)
 			return -DER_CSUM;
 
 		return -DER_TX_UNCERTAIN;
 	}
 
-	memset(obj_auxi->retry_bitmap, 0, OBJ_INLINE_BTIMAP);
-	setbit(obj_auxi->retry_bitmap, next_shard - start_shard);
+	return rc;
+}
+
+static int
+obj_ec_valid_shard_get(struct obj_auxi_args *obj_auxi, uint8_t *tgt_bitmap,
+		       uint32_t grp_idx, uint32_t *tgt_idx)
+{
+	struct dc_object	*obj = obj_auxi->obj;
+	uint32_t		grp_start = grp_idx * obj_get_grp_size(obj);
+	uint32_t		shard_idx = grp_start + *tgt_idx;
+	bool			force_ec_degrade = false;
+	bool			ec_degrade = false;
+	int			rc = 0;
+
+	if (obj_auxi->csum_retry || obj_auxi->tx_uncertain || obj_auxi->force_degraded)
+		force_ec_degrade = true;
```

Comment:
> not sure why for "tx_uncertain" case need to force degraded mode, possibly due to originally modification OP's DTX leader always select parity shard, so if hit TX_UNCERTAIN need to deg-fetch from parity shard?

I thought tx_uncertain is mostly from fetch? instead of modification op. So if normal data shard fetch return uncertain, then do degrade fetch.  Anyway, that is the current master implementation.